egs2/TEMPLATE/s2t1/s2t.sh

```diff
@@ -0,0 +1,1786 @@
+            _frame_shift=
+            if [ -f conf/fbank.conf ] && [ "$(<conf/fbank.conf grep -c frame-shift)" -gt 0 ]; then
+                # Assume using conf/fbank.conf for feature extraction
+                _frame_shift="$(<conf/fbank.conf grep frame-shift | sed -e 's/[-a-z =]*\([0-9]*\)/\1/g')"
+            fi
+            if [ -z "${_frame_shift}" ]; then
+                # If not existing, use the default number in Kaldi (=10ms).
+                # If you are using different number, you have to change the following value manually.
+                _frame_shift=10
+            fi
+
+            _min_length=$(python3 -c "print(int(${min_wav_duration} / ${_frame_shift} * 1000))")
+            _max_length=$(python3 -c "print(int(${max_wav_duration} / ${_frame_shift} * 1000))")
+
+            cp "${data_feats}/org/${dset}/feats_dim" "${data_feats}/${dset}/feats_dim"
+            <"${data_feats}/org/${dset}/feats_shape" awk -F, ' { print $1 } ' \
+                | awk -v min_length="${_min_length}" -v max_length="${_max_length}" \
+                    '{ if ($2 > min_length && $2 < max_length) print $0; }' \
+                    >"${data_feats}/${dset}/feats_shape"
+            <"${data_feats}/org/${dset}/feats.scp" \
+                utils/filter_scp.pl "${data_feats}/${dset}/feats_shape"  \
+                >"${data_feats}/${dset}/feats.scp"
+        fi
+
+        # Remove empty text
+        # shellcheck disable=SC2068
+        for extra_txt in ${utt_extra_files}; do
+            <"${data_feats}/org/${dset}/${extra_txt}" \
+                awk ' { if( NF != 1 ) print $0; } ' >"${data_feats}/${dset}/${extra_txt}"
+        done
+        for ref_txt in ${ref_text_files[@]}; do
+            <"${data_feats}/org/${dset}/${ref_txt}" \
+                awk ' { if( NF != 1 ) print $0; } ' >"${data_feats}/${dset}/${ref_txt}"
+        done
+
+        # fix_data_dir.sh leaves only utts which exist in all files
+        utils/fix_data_dir.sh \
+            --utt_extra_files "${utt_extra_files} ${ref_text_files_str}" \
+            "${data_feats}/${dset}"
+    done
+
+    if [ -n "${post_process_local_data_opts}" ]; then
+        # Do any additional local data post-processing here
+        local/data.sh ${post_process_local_data_opts} --s2t_data_dir "${data_feats}/${train_set}"
+    fi
+
+    # shellcheck disable=SC2002,SC2068,SC2005
+    for lm_txt in ${lm_train_text[@]}; do
+        suffix=$(echo "$(basename ${lm_txt})" | sed 's/text//')
+        <${lm_txt} awk -v suffix=${suffix} ' { if( NF != 1 ) {$1=$1 suffix; print $0; }} '
+    done > "${data_feats}/lm_train.txt"
+fi
+
+
+if [ ${stage} -le 5 ] && [ ${stop_stage} -ge 5 ] && ! [[ " ${skip_stages} " =~ [[:space:]]5[[:space:]] ]]; then
+    if [ "${token_type}" = bpe ]; then
+        log "Stage 5: Generate token_list from ${bpe_train_text} using BPE"
+
+        mkdir -p "${bpedir}"
+        # shellcheck disable=SC2002
+        cat ${bpe_train_text} | cut -f 2- -d" "  > "${bpedir}"/train.txt
```

Comment:
When you use multiple languages, the text data size of different languages are quite different. Would this affect the BPE model? For example, in the extreme case, the rare-resource languages would only have a very small portion in the vocabulary. What do you think of this problem?

Reply:
Yes, this is a problem. Actually in my case, sentencepiece will sample a fixed amount of lines from the entire data (due to memory issue), which can make the situation even worse.

Currently, I don't think our model will work well on low-resource languages due to the imbalanced training data.

We need a better solution as we scale up to more languages.

Reply:
For example, we can just re-use GPT2's vocabulary.

Reply:
I see. Do you think training separate BPE models for different languages and then somehow combining them together is a solution?

Reply:
I'm not sure, since it will make the process very complicated.

I think this is a general machine learning problem. There might be some solutions in NLP.